XLS-39d-clawback/README.md

```diff
@@ -0,0 +1,182 @@
+
+:bangbang: The sub-field `issuer` within `Amount` represents the token holder's address instead of the issuer's.
+
+---
+
+| Field Name     | Required?        |  JSON Type  | Internal Type     |
+|----------------|:----------------:|:-----------:|:-----------------:|
+| `Flags`        |:heavy_check_mark:|`number`     | `UINT32`          |
+
+Specifies the flags for this transaction. The universal transaction flags that are applicable to all transactions (e.g., `tfFullyCanonicalSig`) are valid. This proposal introduces the following new, transaction-specific flags:
+
+>| Flag Name     |         Flag Value         |                                                         Description                                                         |
+>|---------------|:---------------------------------------------------------------------------------------------------------------------------:|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------:|
+>| `tfSetFreeze`    |        `0x00000001`        | If set, either the `lsfHighFreeze` or `lsfLowFreeze` flag (as appropriate) will be set in the trust line. It is not an error to request to set a flag that has already been set. Enabling the `tfSetFreeze` flag simplifies the process of clawing back funds from a trustline. It allows the clawback transaction to freeze the trustline and reclaim the funds in a single transaction, without needing to set the `tfSetFreeze` flag through a `TrustSet` transaction beforehand. However, if the `tfSetFreeze` flag has not been set or the appropriate freeze flag hasn't been enabled on the trustline before attempting to clawback, then the transaction will fail and return the error code `tecNO_PERMISSION`.|
+>| `tfClearFreeze`      |        `0x00000002`        | If set, either the `lsHighFreeze` or `lsfLowFreeze` flag (as appropriate) will be removed from the trust line. It is not an error to request to clear a flag that isn't set. Enabling the `tfClearFreeze` flag simplifies the process of unfreezing a trustline. With this flag enabled, the clawback transaction can reclaim the frozen funds and unfreeze the trustline at the same time, without requiring an additional `TrustSet` transaction to clear the freeze flag.|
+
+The `tfSetFreeze` and `tfClearFreeze` flags are options that can be set in a `Clawback` transaction. They make the process of clawing back funds more convenient by combining multiple steps into one.
+
+Usually, when you want to clawback funds from a trustline, you have to freeze the trustline first, then take back the funds, and finally, unfreeze the trustline if you choose to. With these two flags, you can clawback the funds and keep the trustline unfrozen at the same time by setting both of them simultaneously.
```

Comment:
Clawback and freeze are bundled is because both are meant for regulatory compliance, where freezing a trustline is usually the first step when the issuer notices suspicious activity and then clawback follows. 

Introducing a standalone clawback feature, independent of freeze, can cause concern among holders, particularly they don't know in what ways the issuer will use it. By bundling clawback with freeze, we can provide reassurance for holders that this clawback feature is to be used with the existing freeze feature, which is something they already acknowledge. This also brings a better communication strategy when introducing clawback - we would not be scaring people away by introducing an _standalone_ feature. 

>To me it looks like this completely obviates the need for Clawback to be done on a frozen trust line. It's trivial for the issuer to work around that requirement.

The fact that issuer can easily workaround doesn't conflict with the idea of trying to make users feel safer. I don't see of downside of bundling the clawback with freeze when we can gain a lot benefits, for the reasons stated above. This design choice is made with the best interests of token holders in mind. Yes, issuers may find it silly since they can easily workaround it, but this is okay and not within the consideration of token holders.

Reply:
Your argument for tying clawback to account freezing is marketing based, not engineering based.  I'm an engineer, not a marketing person.  From an engineering perspective, I think tying clawback to account freeze is pointless.  I don't hear you disagreeing with that.

I'm in no sense a marketing person.  So there may be legit marketing reasons to tie those features together.  But I think there is marketing risk as well.  The risk would come in three phases:

1. Some folks promote that clawback is hard for an issuer to abuse because it is tied to trust line freezing.
2. Some (_any_) issuer abuses clawback (easily) because trust line freezing in actuality presents no barrier.
3. Users of the XRP Ledger learn to mistrust those who said that clawback is hard to abuse.

Promoting things that look like safety precautions, but don't actually increase safety, can backfire.  That's worth including in the marketing calculus.

The _real_ safety measures with clawback are two fold:

1. The technical safety measure is that the `lsfAllowClawback` flag must be set on an issuing account before any trust lines are created.  This allows the counterparty of each trust line to decide for themselves whether they trust the issuer not to abuse clawback.
2. The other safety measure is social.  Any issuer that casually performs clawback without strong justification will lose the trust and goodwill of many users of their currency.  Such risks should not be taken lightly.

My belief is that the XRP Ledger should stick to _actual_ safety measures.  Any pretend safety measures risk backfiring.

In summary,

1. I'd prefer that the XRP Ledger not promote fake safety measures.  It's dishonest and can backfire.
2. If we're not going to promote fake safety measures, then let's not include them in our product design.

Reply:
> The _real_ safety measures with clawback are two fold:
> 
> 1. The technical safety measure is that the `lsfAllowClawback` flag must be set on an issuing account before any trust lines are created.  This allows the counterparty of each trust line to decide for themselves whether they trust the issuer not to abuse clawback.
> 2. The other safety measure is social.  Any issuer that casually performs clawback without strong justification will lose the trust and goodwill of many users of their currency.  Such risks should not be taken lightly.

For context, initially, clawback was proposed to be opt-in by default, introducing a certain level of risk, which was why we thought bundling clawback with freeze would be a good idea (so that the issuer can't clawback randomly). Since having clawback enabled by default was really "dangerous", we tied the two features to ensure freeze is a prerequisite to clawback.

Now, with the latest iteration of the spec, clawback is opt-out by default, and has limitation to owner directory, which is a huge safety measure itself. So I do agree that tying clawback with freeze doesn't provide many benefits as before. 

>Your argument for tying clawback to account freezing is marketing based, not engineering based

I don't think we are going to promote the "tying freeze with clawback" as a safety precaution, or do any marketing of that kind. T made sense to do so since both are meant to serve regulatory compliance, hence we engineer the feature around real-world use cases. 

But, I do think it is okay to untie clawback with freeze, and don't have a strong argument to counter that. I'll wait for others to see their opinion.